aesara/tensor/blas.py

```diff
@@ -1022,59 +1033,107 @@ def infer_shape(self, fgraph, node, input_shapes):
+        """
 
-        if (PyArray_DESCR(%(_zout)s)->type_num == NPY_FLOAT)
+    broadcast_xyz = """
+        // Broadcast Z if needed
+        if (Nx[0] > Nz[0] || Ny[1] > Nz[1])
         {
-            float * zoutdata = (float*)PyArray_DATA(%(_zout)s);
-            int zoi = Sz[0] / sizeof(float);
-            int zoj = Sz[1] / sizeof(float);
-            const float * zdata = (float*)PyArray_DATA(%(_z)s);
-            int zi = PyArray_STRIDES(%(_z)s)[0]/sizeof(float);
-            int zj = PyArray_STRIDES(%(_z)s)[1]/sizeof(float);
-            for (int i = 0; i < Nz[0]; ++i)
+            npy_intp dims[2];
+            dims[0] = (PyArray_DIMS(%(_zout)s)[0] >= Nx[0]) ? PyArray_DIMS(%(_zout)s)[0] : Nx[0];
+            dims[1] = (PyArray_DIMS(%(_zout)s)[1] >= Ny[1]) ? PyArray_DIMS(%(_zout)s)[1] : Ny[1];
+            // fprintf(stderr, "Broadcasting Z into shape (%%i %%i)\\n", dims[0], dims[1]);
+            PyArrayObject *z_out_new = (PyArrayObject*)PyArray_SimpleNew(2, dims, PyArray_TYPE(%(_zout)s));
+            if(!z_out_new) {
+                PyErr_SetString(PyExc_MemoryError,
+                                "failed to alloc gemm_inplace output");
+                %(fail)s
+            }
+
+            if(PyArray_MoveInto(z_out_new, %(_zout)s) == -1)
             {
-                for (int j = 0; j < Nz[1]; ++j)
-                {
-                    zoutdata[zoi*i + zoj*j] = zdata[zi*i + zj*j];
-                }
+                %(fail)s
             }
+
+            Py_DECREF(%(_zout)s);
+            %(_zout)s = z_out_new;
+
+            Nz = PyArray_DIMS(%(_zout)s);
+            Sz = PyArray_STRIDES(%(_zout)s);
         }
-        else if (PyArray_DESCR(%(_zout)s)->type_num == NPY_DOUBLE)
+
+        // Broadcast X if needed
+        if (Nz[0] > Nx[0])
         {
-            double * zoutdata = (double*) PyArray_DATA(%(_zout)s);
-            int zoi = Sz[0] / sizeof(double);
-            int zoj = Sz[1] / sizeof(double);
-            const double * zdata = (double*)PyArray_DATA(%(_z)s);
-            int zi = PyArray_STRIDES(%(_z)s)[0]/sizeof(double);
-            int zj = PyArray_STRIDES(%(_z)s)[1]/sizeof(double);
-            for (int i = 0; i < Nz[0]; ++i)
+            npy_intp dims[2];
+            dims[0] = Nz[0];
+            dims[1] = Nx[1];
+            // fprintf(stderr, "Broadcasting X into shape (%%i %%i)\\n", dims[0], dims[1]);
+            PyArrayObject *x_new = (PyArrayObject*)PyArray_SimpleNew(2, dims, PyArray_TYPE(%(_x)s));
+            if(!x_new) {
+                PyErr_SetString(PyExc_MemoryError,
+                                "failed to alloc gemm_inplace input");
+                %(fail)s
+            }
+
+            if(PyArray_MoveInto(x_new, %(_x)s) == -1)
             {
-                for (int j = 0; j < Nz[1]; ++j)
-                {
-                    zoutdata[zoi*i + zoj*j] = zdata[zi*i + zj*j];
-                }
+                %(fail)s
             }
+
+            Py_DECREF(%(_x)s);
```

Comment:
I don't think I want to do this. Instead I want to get rid of the new broadcasted array when I am done.